[FEATURE] Demander à un étudiant de saisir toutes ses informations lors de l'accès à une campagne (PIX-1276).

lisequesnel opened the pull request:
## :unicorn: Problème
Si un étudiant saisi uniquement son numéro étudiant, rien ne nous assure vraiment que c'est bien lui.

## :robot: Solution
On lui demande de saisir quoi qu'il arrive, son numéro étudiant, son prénom, son nom ainsi que sa date de naissance.
On donne accès ou pas selon les règles suivantes :
![image](https://user-images.githubusercontent.com/23451175/93914147-64250400-fd06-11ea-88ad-1a31fc3d7ba4.png)


## :rainbow: Remarques
Scout Rule :
- Un renommage a été effectué sur `HigherEducationRegistration` -> `HigherSchoolingRegistration`
- Des tests manquants sur la méthode `findByOrganizationIdAndStudentNumber` ont été rajoutées

## :100: Pour tester
- Essayer toutes les branches du schéma ci-dessus.

lisequesnel committed:
[SR] Rename reconciliations methods

lisequesnel committed:
[API] Remove usage of reconciliation with only student number

lisequesnel committed:
[APP] Do not make call with only student number

lisequesnel committed:
[API] Validate higher schooling registration reconciliation route

lisequesnel committed:
[SR] Rename findByOrgaIdAndUserData to findBuOrgaIdAndBirthdate

lisequesnel committed:
[SR] Rename HigherEducationRegistration in HigherSchoolingRegistration

lisequesnel committed:
[API] Add constraint on orgaId-studentNumber where isSupernumerary false

lisequesnel committed:
[SR] Rename & Add missing test on findByOrganizationIdAndStudentNumber

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr1906.review.pix.fr
- Orga: https://orga-pr1906.review.pix.fr
- Certif: https://certif-pr1906.review.pix.fr
- Admin: https://admin-pr1906.review.pix.fr
- API: https://api-pr1906.review.pix.fr/api/

 Please check it out!

laura-bergoens commented on api/db/migrations/20200921174050_add-unique-constraint-for-schooling-registrations-on-student-number-and-organization-id-and-not-supernumerary.js:
```diff
@@ -0,0 +1,13 @@
+const TABLE_NAME = 'schooling-registrations';
+
+exports.up = async function(knex) {
+  await knex.schema.table(TABLE_NAME, (table) => table.dropUnique(['studentNumber', 'organizationId']));
+  await knex.raw('CREATE INDEX "organizationid_studentnumber_index" ON "schooling-registrations" ("organizationId", "studentNumber");');
+  return knex.raw('CREATE UNIQUE INDEX "organizationid_studentnumber_notsupernumerary_index" ON "schooling-registrations" ("organizationId", "studentNumber") WHERE "isSupernumerary" IS FALSE;');
+};
+
+exports.down = async function(knex) {
+  await knex.schema.table(TABLE_NAME, (table) => table.unique(['studentNumber', 'organizationId']));
+  await knex.raw('DROP INDEX "organizationid_studentnumber_index" ON "schooling-registrations" ("organizationId", "studentNumber");');
+  return knex.raw('DROP UNIQUE INDEX "organizationid_studentnumber_notsupernumerary_index" ON "schooling-registrations" ("organizationId", "studentNumber") WHERE "isSupernumerary" IS FALSE;');
+};
```
hmmmmm ça va pas je pense. Je sais qu'on en a parlé toutes les deux, mais de le voir je me sens pas bien :face_with_thermometer: 
Il est vrai que la mise en application d'une contrainte d'unicité en PG se traduit (c a d est implémenté) par la création d'un unique index. Mais c'est un détail d'implém, on n'est pas supposé le savoir.

Le SQL de cette ligne là :
```
// js
await knex.schema.table(TABLE_NAME, (table) => table.unique(['studentNumber', 'organizationId']));
// sql
alter table "schooling-registrations" add constraint "schooling_registrations_studentnumber_organizationid_unique" unique ("studentNumber", "organizationId")
```
Ce qui est un peu différent que de mettre l'index à la main. Faut qu'on y réfléchisse encore un peu, voir si on n'a pas d'autres solutions avant de faire celle-ci.

Aussi, quitte à le faire, j'inverserai les lignes (donc je mettrai d'abord les index, puis je drop la contrainte, comme ça pendant la migration pas de risque de casser la cohérence de la base)

lisequesnel replied:
j'ai inversé 👍

ajoanny commented on api/lib/application/schooling-registration-user-associations/schooling-registration-user-association-controller.js:
```diff
@@ -93,15 +92,14 @@ module.exports = {
 
-    const student = {
-      id: request.params.studentId,
+    const higherSchoolingRegistration = {
```
**suggestion (non-blocking) :** On pourrait réutiliser le nommage de `reconciliationInfo` parce que c'est un peut trompeur l'objet n'est pas une `higherSchoolingRegistration`

lisequesnel replied:
C'est pas nom plus une `reconciliationInfo`, c'est ce que saisit l'utilisateur de Pix Orga pour modifier le numéro étudiant

lisequesnel replied:
j'ai passé directement les 2 paramètres au usecase

ajoanny commented on api/lib/domain/services/user-reconciliation-service.js:
```diff
@@ -23,36 +23,51 @@ function findMatchingCandidateIdForGivenUser(matchingUserCandidates, user) {
+  reconciliationInfo: { studentNumber, firstName, lastName, birthdate },
+  schoolingRegistrationRepository,
+}) {
+  const schoolingRegistration = await schoolingRegistrationRepository.findOneImportedByOrganizationIdAndUserData({
```
**suggestion (non-blocking): ** Dans la mesure où s'il n'y a pas de schooling-registrations je lance une erreur j'aurais fait une méthode `getImportedByOrganizationIdAndUserData`.
En faisant ça on s'appuie sur le standard de nommage pour éviter la gestion d'une erreur dans le service en la laissant à la change du repository

ajoanny replied:
Dans le use case `api/lib/domain/usecases/reconcile-higher-schooling-registration.js` on pourrait faire un try catch

lisequesnel replied:
Je pense la logique est différente. Si c'est ok de chercher un schooling registration à partir d'un studentNumber et d'un birthdate et de trouver personne, alors c'est pas une erreur qui est retournée mais juste un résultat vide.
Et c'est notre cas ici. Pour moi le repo doit retourner une erreur si vraiment c'est pas normal. Exemple : on cherche un record à partir de son ID.

ajoanny commented on api/lib/domain/services/user-reconciliation-service.js:
```diff
@@ -23,36 +23,51 @@ function findMatchingCandidateIdForGivenUser(matchingUserCandidates, user) {
+    throw new NotFoundError('There are no schooling registrations found');
+  }
+
+  const schoolingRegistrationId = findMatchingCandidateIdForGivenUser([schoolingRegistration], { firstName, lastName });
+  if (!schoolingRegistrationId) {
+    throw new NotFoundError('There were no schoolingRegistrations matching with names');
+  }
```
**suggestion :**
J'ai l'impression que partout où on utilise `findMatchingCandidateIdForGivenUser` juste derrière on fait :
```js
if (!schoolingRegistrationId) {
    throw new NotFoundError('There were no schoolingRegistrations matching with names');
  }
```
Du  coup on doit pouvoir faire ce throw dans la méthode `findMatchingCandidateIdForGivenUser `

lisequesnel replied:
je propose d'étudier la question et de le faire dans une PR ultérieure si besoin, ça touche à pas mal de choses hors de cette PR.

ajoanny commented on api/lib/domain/usecases/reconcile-higher-schooling-registration.js:
```diff
@@ -0,0 +1,57 @@
+const { NotFoundError } = require('../errors');
+const HigherSchoolingRegistration = require('../models/HigherSchoolingRegistration');
+
+module.exports = async function reconcileHigherSchoolingRegistration({
```
**suggestion (non-blocking):**
J'aurais fait deux usecases :
 - Cas nominal avec le `studentNumber`
 - Cas limite surnuméraire

Après c'est pas très pratique parce que ça implique de créer un nouvelle route, mais comme c'est deux étapes différentes du workflow peut-être que c'est jouable. De cette manière on peut faire évoluer les deux étapes du workflow de manière indépendante

laura-bergoens replied:
On parle de cas limite, en vrai on n'a aucune idée si il est si limite que ça, c'est un peu ce qui me fait peur avec le surnum :grimacing:

lisequesnel replied:
J'ai l'impression qu'on peut pas split en 2 routes, parce qu'au moment où on fait l'appel côté front, on se sait pas dans quel cas on sera.

ajoanny commented on api/lib/domain/usecases/reconcile-higher-schooling-registration.js:
```diff
@@ -0,0 +1,57 @@
+    schoolingRegistrationRepository,
+  });
+
+  return schoolingRegistrationRepository.reconcileUserToSchoolingRegistration({ userId, schoolingRegistrationId: matchedSchoolingRegistration.id });
```
**question :** On a pas besoin de renvoyer de valeur comme c'est une commande. Je sais pas trop c'est quoi le standard et ce vers quoi on veut aller donc c'est parfaitement ok de laisser le return mais je voulais la question.

lisequesnel replied:
pour moi ici, `await` et `return` c'est pareil. On ne récupère aucune valeur, c'est juste pour rendre la main.
J'ai pas de préférence.

ajoanny commented on api/lib/domain/usecases/reconcile-schooling-registration.js:
```diff
@@ -1,11 +1,9 @@
-const Organization = require('../models/Organization');
 
-module.exports = async function reconcileUserToSchoolingRegistrationData({
+module.exports = async function reconcileSchoolingRegistration({
```
**suggestion  (non-blocking):** Le `userTo` dans le nom du use-case apporte un contexte qui est intéressant qui est : avec quoi `schoolinRegistration` est réconcilié.
Par contre le `data` était de trop 👍

bpetetot replied:
Par rapport à la remarque de @ajoanny, je ne pense pas que `userTo`, car on ne peut pas reconcilier autre chose qu'un utilisateur. La réconcialiation c'est toujours `User <=> Schooling Registration || HigherSchoolingRegistration`.

ajoanny commented on api/lib/domain/usecases/update-student-number.js:
```diff
@@ -2,18 +2,18 @@ const _ = require('lodash');
+  const { id, studentNumber } = higherSchoolingRegistration;
 
-  const [schoolingRegistration] = await schoolingRegistrationRepository.findOneByOrganizationIdAndStudentNumber(organizationId, studentNumber);
+  const foundHigherSchoolingRegistrations = await schoolingRegistrationRepository.findByOrganizationIdAndStudentNumber({ organizationId, studentNumber });
```
**suggestion (non-bloking):** Le mauvais maladroit du paramètre `higherSchoolingRegistration` force l'utilisation du nom `foundHigherSchoolingRegistrations` ce qui n'est pas un nom très intéressant. 
Le paramètre `higherSchoolingRegistrations`à un nom trompeur puisque ce n'est pas une `HigherSchoolingRegistrations` peut-être que les passer en deux paramètres serait mieux dans ce cas

lisequesnel replied:
ok pour le passer en 2 paramètres

ajoanny commented on api/lib/infrastructure/repositories/schooling-registration-repository.js:
```diff
@@ -107,19 +107,18 @@ module.exports = {
   },
 
-  async findOneByOrganizationIdAndStudentNumber(organizationId, studentNumber) {
+  async findByOrganizationIdAndStudentNumber({ organizationId, studentNumber }) {
```
**suggestion (non-blocking) :** Je pense que cette méthode aurait plus sa place dans le `HigherSchoolingRegistrationRepository`

ajoanny replied:
Pareil pour `findOneImportedByOrganizationIdAndUserData`

lisequesnel replied:
ah j'avais pas tilté qu'il existait, je les bouge

ajoanny commented on api/tests/integration/application/schooling-registration-user-associations/index_test.js:
```diff
@@ -219,12 +228,130 @@ describe('Integration | Application | Route | schooling-registration-user-associ
       });
 
-      it('should failed if student number is an empty string', async () => {
+      it('should succeed when there is a space', async () => {
```
**question :** J'ai l'impression que c'est des tests de validations qu'on devrait avoir dans les validations le modèle ? Je suis pas sûr de comprendre l'intérêt de le tester içi ? C'est à cause des validations Joi sur la route ?

bpetetot replied:
Oui c'est la validation Joi, notamment la partie `regexp`
```js
Joi.string().empty(Joi.string().regex(/^\s*$/))
```

ajoanny commented on api/tests/integration/domain/usecases/reconcile-higher-schooling-registration_test.js:
```diff
@@ -0,0 +1,221 @@
+
+    context('When no imported schooling registration found with student number', () => {
+      context('When student number already exists in supernumerary', () => {
+        it('should save supernumerary schooling registration with user info', async () => {
```
**suggestion :** A la lecture on pourrait croire qu'on  met ajout la schooling registration existante mais en  faite on créer un  nouveau record je crois (insert) ?

ajoanny commented on api/tests/integration/domain/usecases/reconcile-higher-schooling-registration_test.js:
```diff
@@ -0,0 +1,221 @@
+      });
+
+      context('When student number does not already exist in supernumerary', () => {
+        it('should save supernumerary schooling registration with user info', async () => {
```
**suggestion :**  L'utilisation du mot `save` dans les descriptions me fait penser qu'on fait des updates. Tu penses quoi @lisequesnel ?

lisequesnel replied:
je peux changer pour être plus explicite (`update` et `create`)

ajoanny commented on api/tests/integration/infrastructure/repositories/schooling-registration-repository_test.js:
```diff
@@ -512,153 +512,82 @@ describe('Integration | Infrastructure | Repository | schooling-registration-rep
-      it('should return empty array with wrong studentNumber', async () => {
-        // given
-        const reconciliationInfo = { studentNumber: '789B' };
+    it('should return empty array with wrong birthdate', async () => {
```
**suggestion :**
```js
    it('should return empty array when there is no schooling-registrations with the  given birthdate', async () => {

```

ajoanny commented on api/tests/integration/infrastructure/repositories/schooling-registration-repository_test.js:
```diff
@@ -512,153 +512,82 @@ describe('Integration | Infrastructure | Repository | schooling-registration-rep
+      expect(result.length).to.be.equal(0);
     });
 
     it('should return empty array with fake organizationId', async () => {
```
**suggestion :**
```js
    it('should return empty array when there is no schooling-registrations with the  given organizationId', async () => {
```

ajoanny commented on api/tests/integration/infrastructure/repositories/schooling-registration-repository_test.js:
```diff
@@ -512,153 +512,82 @@ describe('Integration | Infrastructure | Repository | schooling-registration-rep
-        const result = await schoolingRegistrationRepository.findSupernumeraryByOrganizationIdAndBirthdate({
-          organizationId: '999', birthdate,
-        });
+    it('should return empty array with wrong student number', async () => {
```
**suggestion :**
Je pense la description `with wrong XXXX` ou `with  fake  XXXX` n'est pas correct dans  ces  test on ne décrit pas le comportement du code

ajoanny commented on api/tests/integration/infrastructure/repositories/schooling-registration-repository_test.js:
```diff
@@ -1229,4 +1158,127 @@ describe('Integration | Infrastructure | Repository | schooling-registration-rep
+      });
+    });
+
+    context('When there is no schooling registrations for the organization', () => {
```
**suggestion :** Le  beforeEach ligne 1167 crée un schooling registration  pour l'organisation

lisequesnel replied:
il créé juste une organization

ajoanny commented on api/tests/integration/infrastructure/repositories/schooling-registration-repository_test.js:
```diff
@@ -1229,4 +1158,127 @@ describe('Integration | Infrastructure | Repository | schooling-registration-rep
+      });
+    });
+
+    context('When there is no schooling registrations with given student number', () => {
```
**suggestion :** Il manque le contexte `when there are imported schooling registrations`  Là on est dans le cas où il  y a  de  des `schooling-registrations`

lisequesnel replied:
Rajouter ce contexte serait moins clair je trouve, sachant que les autres tests couvrent déjà les différents cas

ajoanny commented on api/lib/infrastructure/repositories/schooling-registration-repository.js:
```diff
@@ -130,16 +129,17 @@ module.exports = {
 
-  async findSupernumeraryByOrganizationIdAndBirthdate({ organizationId, birthdate }) {
-    const schoolingRegistrations = await BookshelfSchoolingRegistration
+  async findOneImportedByOrganizationIdAndUserData({ organizationId, reconciliationInfo: { birthdate, studentNumber } = {} }) {
```
**suggestion :** Le mot `imported` fait un peu désinformation parce que l'idée derrière c'est que c'est sur-numeraire  `importées` c'est juste un manière d'avoir des schooling-registration non surnuméraire.
A la fin les `schooling-registration` surnuméraire vont être validé et elle ne seront plus surnuméraire mais n'auront pas été importées.

lisequesnel replied:
Tu proposes quel wording en remplacement ?

lisequesnel replied:
`registered` ?

ajoanny commented on api/lib/infrastructure/repositories/schooling-registration-repository.js:
```diff
@@ -130,16 +129,17 @@ module.exports = {
-        qb.where('birthdate', birthdate);
-        qb.where('isSupernumerary', true);
+        qb.where('isSupernumerary', false);
+        if (birthdate) qb.where('birthdate', birthdate);
+        if (studentNumber) qb.whereRaw('LOWER(?)=LOWER(??)', [studentNumber, 'studentNumber']);
```
**suggestion : ** Je pense que c'est un mauvais parttern de faire ça. On factorise  du code avec deux intention  différente. Soit on veut   par `birthdate` soit on veut par `studentNumber`.
Avec ces paramètres on à 4 combinaison possible  si on avait  deux  méthodes on en aurait que 2.

bpetetot replied:
De mémoire, il me semble qu'on a besoin des 3 cas:
- uniquement birthdate
- uniquement student number
- birthdate + student number

Du coup, cette petite factorisation intéressante. Mais peut être que la fonction est male nommée.

lisequesnel replied:
Une proposition de naming ?

ajoanny commented on api/tests/unit/domain/services/user-reconciliation-service_test.js:
```diff
@@ -364,78 +364,121 @@ describe('Unit | Service | user-reconciliation-service', () => {
+
+        context('When schoolingRegistration is not already linked', () => {
+
+          it('should return matchedSchoolingRegistration', async () => {
```
**suggestion :** 
```js
          it('should return matched SchoolingRegistration', async () => {
```

ajoanny commented on mon-pix/app/components/routes/campaigns/restricted/join-sup.hbs:
```diff
@@ -7,15 +7,15 @@
       {{this.validation.studentNumber}}
     </div>
-    {{#if this.showSupernumeraryForm}}
+    {{#if this.showFurtherInformationForm}}
```
**question: ** Pourquoi on garde pas l'idée de supernumerary ? Le terme n'est pas terrible ok mais c'est plus spécifique que `showFurtherInformationForm` qui est très générique et qu'on pourrait retrouver dans n'importe quel formulaire.
Je pense que `showSupernumeraryForm` passe plus clairement l'intention, mais c'est peut-être pour cacher cette notion dans le front ?

lisequesnel replied:
Parce que maintenant, on montre ce formulaire à tout le monde, qu'on les inscrire en surnuméraire ou qu'on les trouve dans les inscrits.

ajoanny commented on mon-pix/app/components/routes/campaigns/restricted/join-sup.js:
```diff
@@ -139,25 +134,18 @@ export default class JoinSup extends Component {
-      if (error.status === '404') {
-        this.showSupernumeraryForm = true;
-        return this.errorMessageForStudentNumberForm = 'Vérifiez votre numéro étudiant ou saisissez les informations ci-dessous.';
+      if (error.status === '409' || error.status === '404') {
```
**suggestion (non-blocking) :**
On pourrait extraire les magic value dans des constantes

```js
if(ERRORS_HANDLED.includes(error.status){
  return this.errorMessage = ERRORS_HANDLED_MESSAGE;
}
```

Je trouve pas très propre de faire un `return` alors qu'on utilise pas la valeur de retour de la fonction.

lisequesnel replied:
De mémoire, avec Ember, si tu `return` pas, Ember peut rester bloqué et pas avoir le comportement que tu souhaites

ajoanny commented on mon-pix/tests/integration/components/routes/campaigns/restricted/join-sup-test.js:
```diff
@@ -40,42 +37,8 @@ describe('Integration | Component | routes/campaigns/restricted/join-sup', funct
-      expect(onSubmitToReconcileStub.getCall(1).args[0].toJSON()).to.deep.equal(schoolingRegistration);
-    });
-
     it('should disable input student number', async function() {
```
**suggestion :** Le contexte de ce test est faux je crois ?
Quand le `studentNumber` est tapé ça ne désactive pas l'input.

```js
  context('when the student number is typed', () => {
   it('should disable input student number',
```
Je vais regarder sur la RA

lisequesnel replied:
si ça grise toujours l'input

ajoanny commented on mon-pix/tests/integration/components/routes/campaigns/restricted/join-sup-test.js:
```diff
@@ -91,7 +54,7 @@ describe('Integration | Component | routes/campaigns/restricted/join-sup', funct
 
   context('when i want change the student number', () => {
-    it('should be possible de edit student number', async function() {
+    it('should be possible to edit student number', async function() {
```
**question :**
Je pense qu'on peut rajouter dans le contexte la raison pour laquelle lil est possible de changer le `studentNumber`

ajoanny commented on api/tests/integration/domain/usecases/reconcile-higher-schooling-registration_test.js:
```diff
@@ -0,0 +1,221 @@
+      });
+    });
+
+    context('When no imported schooling registration found with matching student number, firstName, lastName and birthdate', () => {
```
**suggestion :**
Le mot `imported` n'est pas le plus approprié pour décrire ce cas

ajoanny commented on api/tests/integration/domain/usecases/reconcile-higher-schooling-registration_test.js:
```diff
@@ -0,0 +1,221 @@
+
+const reconcileHigherSchoolingRegistration = require('../../../../lib/domain/usecases/reconcile-higher-schooling-registration');
+
+describe('Integration | UseCases | reconcile-higher-schooling-registration', () => {
```
**suggestion (non-blocking) :**
Je trouve que parcourir les tests n'est pas simple parce qu'il manque des contextes.
Exemple :
il y a `context('When there is no student number', () => {` mais il n'y a pas l'inverse du coup j'ai du mal à comprendre dans quel  cas je suis. C'est peut-être une question d'habitude aussi

lisequesnel replied:
L'imbrication des contextes complexifie aussi la lecture. C'est un peu au jugement de chaque personne j'ai l'impression.

lisequesnel committed:
[API] Reconcile or save supernumerary higher schooling registrations

lisequesnel committed:
[SR] Remove findSupernumeraryByOrganizationIdAndBirthdate

lisequesnel committed:
[SR] Move 3 methods into higherSchoolingRegistrationRepository

- findOneImportedByOrganizationIdAndUserData
- findByOrganizationIdAndStudentNumber
- updateStudentNumber

ajoanny commented:
@lisequesnel Tests fonctionnels : 👍 !
